graphula-core/src/Graphula.hs

```diff
@@ -429,85 +467,234 @@ class HasDependencies a where
+    ':<>: 'Text "'SourceDefault"
+    ':<>: 'Text " or "
+    ':<>: 'Text "'SourceArbitrary"
+  ) => GenEntityKey 'SourceExternal a where
+  type KeyConstraint 'SourceExternal a = NoConstraint a
+  genEntityKey = error "unreachable"
+
+data GenerationFailure
+  = GenerationFailureMaxAttemptsToConstrain TypeRep
+  -- ^ Could not satisfy constraints defined using @'ensure'@
+  | GenerationFailureMaxAttemptsToInsert TypeRep
+  -- ^ Could not satisfy database constraints on insert
   deriving (Show, Typeable, Eq)
 
 instance Exception GenerationFailure
 
-type GraphulaNode m a =
-  ( Generate m a
-  , HasDependencies a
-  , Logging m a
-  , PersistEntityBackend a ~ SqlBackend
-  , PersistEntity a
-  , Typeable a
-  , EntityKeyGen a
-  )
+type GraphulaNode m a
+  = ( Generate m a
+    , HasDependencies a
+    , Logging m a
+    , PersistEntityBackend a ~ SqlBackend
+    , PersistEntity a
+    , Typeable a
+    )
 
 {-|
-  Generate and edit a value with data dependencies. This leverages
+  Generate a value with data dependencies. This leverages
   'HasDependencies' to insert the specified data in the generated value. All
-  dependency data is inserted after editing operations.
+  dependency data is inserted after any editing operations.
 
-  > nodeEdit @Dog (ownerId, veterinarianId) $ \dog ->
+  > node @Dog (ownerId, veterinarianId) mempty
+  > node @Dog (ownerId, veterinarianId) $ edit $ \dog ->
   >   dog {name = "fido"}
+
+  A value that has an externally managed key must use @'nodeKeyed'@ instead.
 -}
-nodeEditWith
-  :: forall a m
-   . GraphulaContext m '[a]
+node
+  :: forall a s m
+   . ( GraphulaContext m '[a]
+     , KeySource a ~ s
+     , GenEntityKey s a
+     , KeyConstraint s a
+     )
   => Dependencies a
-  -> (a -> a)
-  -> m (Entity a)
-nodeEditWith dependencies edits = 10 `attemptsToInsertWith` do
-  x <- (`dependsOn` dependencies) . edits <$> generateNode
-  logNode x
-  pure x
+  -> NodeOptions a
+  -> (m (Entity a))
+node = nodeImpl $ genEntityKey @s @a
 
 {-|
-  Generate a value with data dependencies. This leverages 'HasDependencies' to
-  insert the specified data in the generated value.
-
-  > nodeEdit @Dog (ownerId, veterinarianId)
+  Generate a value with data dependencies given an externally managed
+  key. This leverages 'HasDependencies' to insert the specified data
+  in the generated value. All dependency data is inserted after any
+  editing operations.
+
+  > someKey <- generateKey
+  > node @Cat someKey (ownerId, veterinarianId) mempty
+  > anotherKey <- generateKey
+  > node @Cat anotherKey (ownerId, veterinarianId) $ edit $ \cat ->
+  >   cat {name = "milo"}
+
+  A value that has an automatically managed key may use @'node'@ instead.
 -}
-nodeWith
-  :: forall a m . GraphulaContext m '[a] => Dependencies a -> m (Entity a)
-nodeWith = flip nodeEditWith id
-
-{-|
-  Generate and edit a value that does not have any dependencies.
+nodeKeyed
+  :: forall a m
+   . GraphulaContext m '[a]
+  => Key a
```

Comment:
If I generate a `SourceDefault` or `SourceArbitrary` this will not throw a type error?

Reply:
Yes, I currently allow you to specify a key for those types, e.g.:

```haskell
instance HasDependencies D where
  type KeySource D = 'SourceArbitrary

deriving instance {-# OVERLAPPING #-} Arbitrary (Key D)

...
  k <- liftIO $ generate arbitrary
  _ <- nodeKeyed @D k () mempty
```

^^^ that one is probably harmless. This is probably a mistake:

```haskell
instance HasDependencies A where
  KeySourceA = 'SourceDefault

let k = AKey 1
_ <- nodeKeyed @A k () mempty
```

Tho, worst case scenario, you get a gen-failure when you re-run the test.

It'd be easy to disallow both cases:

```haskell
nodeKeyed
  :: forall a m
   . ( GraphulaContext m '[a]
     , KeySource a ~ 'SourceExternal
     )
  => Dependencies a
  -> NodeOptions a
  -> (m (Entity a))
```

If we just want to disallow the `'SourceExternal` case but allow `'SourceArbitrary`, it's a little more work.

Reply:
Mm, even the `'SourceArbitrary` case is only safe for `nodeKeyed` if your `Arbitrary` instance always returns a unique key (e.g. `UUID`), so maybe we should disallow both.

Reply:
I don't think there is anything in principle with allowing `SourceArbitrary` and `SourceDefault` from using `keyed`. My open question is whether those uses are more likely to be intentional or errors. :man_shrugging: I think I'm fine with allowing it and seeing how it shakes out.

Reply:
Yep, same